packages/js/product-editor/src/hooks/use-product-template/use-product-template.ts

```diff
@@ -38,10 +52,35 @@ export const useProductTemplate = (
 	);
 
 	if ( ! matchingProductTemplate ) {
-		// Fallback to the first template with the same product type.
-		matchingProductTemplate = productTemplates.find(
+		// Look for matching templates based on product data described on each template.
+		const matchingTemplates = productTemplates.filter(
 			( productTemplate ) =>
-				productTemplate.productData.type === productTypeToFind
+				Object.entries( productTemplate.productData ).every(
+					( [ key, value ] ) => {
+						if ( ! product ) {
+							return false;
+						}
+
+						if ( key === 'meta_data' ) {
+							return matchesAllTemplateMetaFields(
+								value,
+								( product && product[ key ] ) || []
+							);
+						}
+
+						return product[ key ] === value;
+					}
+				)
```

Comment:
I'd extract this function to a separate one called something like `matchesAllTemplateProductData()`